src/profile.c

```diff
@@ -487,13 +483,61 @@ static inline void printOutput(UInt line, int nameid, int exec, int visited)
     }
 }
 
+// Mark line as visited, and return true if the line has been previously
+// visited (executed)
+BOOL markVisited(int nameid, UInt line)
+{
+    if (nameid == 0 || line == 0) {
+        return 1;
+    }
+
+    if (LEN_PLIST(profileState.visitedStatements) < nameid ||
+        !ELM_PLIST(profileState.visitedStatements, nameid)) {
+        AssPlist(profileState.visitedStatements, nameid,
+                 NEW_PLIST(T_PLIST, 0));
+    }
+
+    Obj linelist = ELM_PLIST(profileState.visitedStatements, nameid);
+
+    if (LEN_PLIST(linelist) < line || !ELM_PLIST(linelist, line)) {
+        AssPlist(linelist, line, True);
```

Comment:
Perhaps we can use a blist instead of a plist to reduce memory usage by a factor of 64 or so? I think the main missing ingredient would be a function which takes a blist and efficiently increases its size by adding zero blocks.

Reply:
Oh and perhaps another helper which takes a blist and an index and returns `TRUE` or `FALSE` (not `True` or `False`) ... this helper then could also just return 0 for anything beyond the bounds. That would simplify the code here a bit, no?

Reply:
As you say, I started trying to use blists then discovered there is no way to extend a blist while having it remain a blist.

It probably wouldn't be a bad idea to add such extensions to blists (maybe even accessible from the GAP level), but that was a rabbit-hole I didn't want to fall down, as it could be done later / more generally (should we just add a generic "extend list, and use this member to fill the new values, and special case that for blists?)

Reply:
I'll have a quick look. I worry about not testing such functions (you didn't do SET_LEN_BLIST in SAFE_SET_BIT_BLIST, which I think is required), I will take a look at doing this with BLists, however.

Reply:
Oh sure. I'm not even sure that code compiles, I just copy&pasted it quickly. Yeah, it should have something like `if (pos > LEN_BLIST(list)) SET_LEN_BLIST(list, pos);` added